### PR TITLE
[docs][linstor] Backups on S3 storage

### PR DESCRIPTION
**Pull request by z9r5**
## Description
Describes how to backup linstor volume to S3 storage.

Ref: #1503 

Depends on https://github.com/deckhouse/deckhouse/pull/2054

## Changelog entries
```changes
section: docs
type: feature
summary: Described how to backup linstor volume to S3 storage.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 1]
```diff
@@ -98,3 +98,222 @@ In case your application does not support high availability and runs in a single
 To solve the problem, specify the label `linstor.csi.linbit.com/on-storage-lost: remove` in the Pod description. The linstor module will automatically remove such Pods from the node where the problem occurred, allowing Kubernetes to restart the application on another node.
 
 [Example...](usage.html#application-reschedule-in-case-of-node-problem-storage-based-fencing)
+
+## Backup on S3 Storage
+
+Data backup is implemented using [volume snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Snapshots are supported by the [snapshot-controller](../045-snapshot-controller/) module, which is enabled automatically for supported CSI drivers in Kubernetes clusters versions 1.20 and higher.
+
+> It works with [Amazon S3](https://aws.amazon.com/s3/) and [MinIO](min.io). There may be problems when working with other services.
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:106]: link does not work
```suggestion
> It works with [Amazon S3](https://aws.amazon.com/s3/) and [MinIO](https://min.io). There may be problems when working with other services.  
```

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 2]
```diff
@@ -98,3 +98,222 @@ In case your application does not support high availability and runs in a single
...
 
 [Example...](usage.html#application-reschedule-in-case-of-node-problem-storage-based-fencing)
+
+## Backup on S3 Storage
+
+Data backup is implemented using [volume snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Snapshots are supported by the [snapshot-controller](../045-snapshot-controller/) module, which is enabled automatically for supported CSI drivers in Kubernetes clusters versions 1.20 and higher.
+
+> It works with [Amazon S3](https://aws.amazon.com/s3/) and [MinIO](min.io). There may be problems when working with other services.  
+>
+> Snapshot are supported only for LVMThin pools.
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:108]:
```suggestion
> Snapshots are supported only for LVMThin pools.
```

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 3]
```diff
@@ -98,3 +98,222 @@ In case your application does not support high availability and runs in a single
...
+   apiVersion: snapshot.storage.k8s.io/v1
+   kind: VolumeSnapshot
+   metadata:
+     name: example-backup-from-s3
+     namespace: storage
+   spec:
+     source:
+       volumeSnapshotContentName: restored-snap-content-from-s3 # The name of the VolumeSnapshotContent created earlier.
+     volumeSnapshotClassName: linstor-csi-snapshot-class-s3     # The name of the VolumeSnapshotClass, with access to the S3 storage.
+   ```
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:294]: This is needed only for recovering into another cluster/namespace, isn't it?

File: modules/031-linstor/docs/CONFIGURATION_RU.md [hunk 4]
```diff
@@ -98,3 +98,222 @@ data:
 Для решения этой задачи укажите лейбл `linstor.csi.linbit.com/on-storage-lost: remove` в описании Pod'а. Модуль linstor автоматически удалит такие Pod'ы с узла где возникли проблемы, что позволит Kubernetes перезапустить приложение на другом узле.
 
 [Пример...](usage.html#перенос-приложения-при-проблемах-с-узлом-storage-based-fencing)
+
+## Резервное копирование в S3
+
+Резервное копирование данных реализовано с помощью [снапшотов томов](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Поддержка работы снапшотов обеспечивается модулем [snapshot-controller](../045-snapshot-controller/), который включается автоматически для поддерживаемых CSI-драйверов в кластерах Kubernetes версий 1.20 и выше.
+
+> Возможна работа не со всеми S3-совместимыми сервисами. Подтверждена работа с [Amazon S3](https://aws.amazon.com/s3/) и [MinIO](min.io).
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION_RU.md:106]:
```suggestion
> Возможна работа не со всеми S3-совместимыми сервисами. Подтверждена работа с [Amazon S3](https://aws.amazon.com/s3/) и [MinIO](https://min.io).
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

File: modules/031-linstor/docs/CONFIGURATION.md (hunk 3 above)
Comment (z9r5) [reply modules/031-linstor/docs/CONFIGURATION.md]: Made the changes. Please take a look on a [stage](https://deckhouse.stage.flant.com/en/documentation/v1/modules/031-linstor/configuration.html#restoring-from-a-backup) @kvaps

File: modules/031-linstor/docs/CONFIGURATION_RU.md [hunk 5]
```diff
@@ -98,3 +98,249 @@ data:
 Для решения этой задачи укажите лейбл `linstor.csi.linbit.com/on-storage-lost: remove` в описании Pod'а. Модуль linstor автоматически удалит такие Pod'ы с узла где возникли проблемы, что позволит Kubernetes перезапустить приложение на другом узле.
 
 [Пример...](usage.html#перенос-приложения-при-проблемах-с-узлом-storage-based-fencing)
+
+## Резервное копирование в S3
+
+Резервное копирование данных реализовано с помощью [снапшотов томов](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Поддержка работы снапшотов обеспечивается модулем [snapshot-controller](../045-snapshot-controller/), который включается автоматически для поддерживаемых CSI-драйверов в кластерах Kubernetes версий 1.20 и выше.
+
+> Возможна работа не со всеми S3-совместимыми сервисами. Подтверждена работа с [Amazon S3](https://aws.amazon.com/s3/) и [MinIO](https://min.io).
+>
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION_RU.md:107]:
```suggestion
```

fixed in [v1.19.0](https://github.com/LINBIT/linstor-server/releases/tag/v1.19.0)

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 6]
```diff
@@ -98,3 +98,249 @@ In case your application does not support high availability and runs in a single
 To solve the problem, specify the label `linstor.csi.linbit.com/on-storage-lost: remove` in the Pod description. The linstor module will automatically remove such Pods from the node where the problem occurred, allowing Kubernetes to restart the application on another node.
 
 [Example...](usage.html#application-reschedule-in-case-of-node-problem-storage-based-fencing)
+
+## Backup on S3 Storage
+
+Data backup is implemented using [volume snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Snapshots are supported by the [snapshot-controller](../045-snapshot-controller/) module, which is enabled automatically for supported CSI drivers in Kubernetes clusters versions 1.20 and higher.
+
+> It works with [Amazon S3](https://aws.amazon.com/s3/) and [MinIO](https://min.io). There may be problems when working with other services.  
+>
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:107]:
```suggestion
```

fixed in [v1.19.0](https://github.com/LINBIT/linstor-server/releases/tag/v1.19.0)

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 7]
```diff
@@ -98,3 +98,249 @@ In case your application does not support high availability and runs in a single
 To solve the problem, specify the label `linstor.csi.linbit.com/on-storage-lost: remove` in the Pod description. The linstor module will automatically remove such Pods from the node where the problem occurred, allowing Kubernetes to restart the application on another node.
 
 [Example...](usage.html#application-reschedule-in-case-of-node-problem-storage-based-fencing)
+
+## Backup on S3 Storage
+
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:103]:
```suggestion

> Использование данной возможности требует настроенного мастер-пароля

```

ref https://github.com/deckhouse/deckhouse/pull/2054

File: modules/031-linstor/docs/CONFIGURATION_RU.md [hunk 8]
```diff
@@ -98,3 +98,249 @@ data:
 Для решения этой задачи укажите лейбл `linstor.csi.linbit.com/on-storage-lost: remove` в описании Pod'а. Модуль linstor автоматически удалит такие Pod'ы с узла где возникли проблемы, что позволит Kubernetes перезапустить приложение на другом узле.
 
 [Пример...](usage.html#перенос-приложения-при-проблемах-с-узлом-storage-based-fencing)
+
+## Резервное копирование в S3
+
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION_RU.md:103]:
```suggestion

> This feature requires a configured master passphrase.

```

ref https://github.com/deckhouse/deckhouse/pull/2054

File: modules/031-linstor/docs/CONFIGURATION_RU.md [hunk 9]
```diff
@@ -98,3 +98,249 @@ data:
...
+   apiVersion: v1
+   metadata:
+     name: linstor-csi-s3-access
+     namespace: storage
+   immutable: true
+   type: linstor.csi.linbit.com/s3-credentials.v1
+   stringData:
+     access-key: *!ACCESS_KEY*  # Access key доступа к хранилищу S3.
+     secret-key: *!SECRET_KEY*  # Secret key доступа к хранилищу S3.
+   ```
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION_RU.md:169]:
```suggestion
Для создания снапшота тома и загрузки его в S3 выполните следующие шаги:
1. Создайте `VolumeSnapshotClass` и `Secret`, содержащий access key и secret key доступа к хранилищу S3.

   > VolumeSnapshotClass — ресурс на уровне кластера. Один и тот же VolumeSnapshotClass можно использовать для создания резервных копий разных PVC из разных пространств имен.

   Пример `VolumeSnapshotClass` и `Secret`:

   ```yaml
   kind: VolumeSnapshotClass
   apiVersion: snapshot.storage.k8s.io/v1
   metadata:
     name: linstor-csi-snapshot-class-s3
   driver: linstor.csi.linbit.com
   deletionPolicy: Retain
   parameters:
     snap.linstor.csi.linbit.com/type: S3
     snap.linstor.csi.linbit.com/remote-name: backup-remote               # Уникальное название backup-подключения в linstor.   
     snap.linstor.csi.linbit.com/allow-incremental: "false"               # Использовать ли инкрементальные копии. 
     snap.linstor.csi.linbit.com/s3-bucket: snapshot-bucket               # Название S3 bucket, для хранения данных.
     snap.linstor.csi.linbit.com/s3-endpoint: s3.us-west-1.amazonaws.com  # S3 endpoint URL.
     snap.linstor.csi.linbit.com/s3-signing-region: us-west-1             # Регион S3. 
     # Использовать virtual hosted–style или path-style S3 URL 
     # https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html
     snap.linstor.csi.linbit.com/s3-use-path-style: "false"    
     # Ссылка на Secret, содержащий access key и secret key доступа к S3 bucket.
     csi.storage.k8s.io/snapshotter-secret-name: linstor-csi-s3-access
     csi.storage.k8s.io/snapshotter-secret-namespace: storage
   ---
   kind: Secret
   apiVersion: v1
   metadata:
     name: linstor-csi-s3-access
     namespace: storage
   immutable: true
   type: linstor.csi.linbit.com/s3-credentials.v1
   stringData:
     access-key: *!ACCESS_KEY*  # Access key доступа к хранилищу S3.
     secret-key: *!SECRET_KEY*  # Secret key доступа к хранилищу S3.
   ```

1. Выберите (или создайте) `PersistentVolumeClaim`, данные которого нужно копировать.

   Пример `PersistentVolumeClaim`, который будет использоваться в примерах далее:

   ```yaml
   apiVersion: v1
   kind: PersistentVolumeClaim
   metadata:
     name: my-linstor-volume
     namespace: storage
   spec:
     accessModes:
     - ReadWriteOnce
     storageClassName: linstor-thindata-r2   # StorageClass хранилища linstor.
     resources:
       requests:
         storage: 2Gi
   ```
```

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 10]
```diff
@@ -98,3 +98,249 @@ In case your application does not support high availability and runs in a single
...
+   apiVersion: v1
+   metadata:
+     name: linstor-csi-s3-access
+     namespace: storage
+   immutable: true
+   type: linstor.csi.linbit.com/s3-credentials.v1
+   stringData:
+     access-key: *!ACCESS_KEY*  # S3 access key.
+     secret-key: *!SECRET_KEY*  # S3 secret key.
+   ```
```
Comment (kvaps) [inline modules/031-linstor/docs/CONFIGURATION.md:169]:
```suggestion
To create a snapshot of a volume and upload it to S3, follow these steps:
1. Create `VolumeSnapshotClass` and `Secret`, containing the access key and secret key of S3 storage.

   > VolumeSnapshotClass is a cluster-wide resource. The same VolumeSnapshotClass can be used to create backups of different PVCs from different namespaces.

   Example of `VolumeSnapshotClass` and `Secret`:

   ```yaml
   kind: VolumeSnapshotClass
   apiVersion: snapshot.storage.k8s.io/v1
   metadata:
     name: linstor-csi-snapshot-class-s3
   driver: linstor.csi.linbit.com
   deletionPolicy: Retain
   parameters:
     snap.linstor.csi.linbit.com/type: S3
     snap.linstor.csi.linbit.com/remote-name: backup-remote               # Linstor remote name.   
     snap.linstor.csi.linbit.com/allow-incremental: "false"               # Whether to use incremental copies. 
     snap.linstor.csi.linbit.com/s3-bucket: snapshot-bucket               # The name of the S3 bucket, for data storage.
     snap.linstor.csi.linbit.com/s3-endpoint: s3.us-west-1.amazonaws.com  # S3 endpoint URL.
     snap.linstor.csi.linbit.com/s3-signing-region: us-west-1             # The name of the S3 region. 
     # Use virtual hosted–style or path-style S3 URL. 
     # https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html
     snap.linstor.csi.linbit.com/s3-use-path-style: "false"    
     # Secret, containing the access key and secret key of access to S3 bucket.
     csi.storage.k8s.io/snapshotter-secret-name: linstor-csi-s3-access
     csi.storage.k8s.io/snapshotter-secret-namespace: storage
   ---
   kind: Secret
   apiVersion: v1
   metadata:
     name: linstor-csi-s3-access
     namespace: storage
   immutable: true
   type: linstor.csi.linbit.com/s3-credentials.v1
   stringData:
     access-key: *!ACCESS_KEY*  # S3 access key.
     secret-key: *!SECRET_KEY*  # S3 secret key.
   ```

1. Select (or create) `PersistentVolumeClaim`, whose data needs to be copied.

   Example of `PersistentVolumeClaim`, which will be used in the examples below:

   ```yaml
   apiVersion: v1
   kind: PersistentVolumeClaim
   metadata:
     name: my-linstor-volume
     namespace: storage
   spec:
     accessModes:
     - ReadWriteOnce
     storageClassName: linstor-thindata-r2   # The name of the StorageClass.
     resources:
       requests:
         storage: 2Gi
   ```
```

**Review by kvaps — COMMENTED** [review]
I would like to swap sections 1. and 2. for creating backup

**Comment by kvaps** [comment]
Also shouldn't we move such information into USAGE.md ?

File: modules/031-linstor/docs/CONFIGURATION_RU.md (hunk 9 above)
Comment (z9r5) [reply modules/031-linstor/docs/CONFIGURATION_RU.md]: Fixed

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

File: modules/031-linstor/docs/USAGE.md [hunk 11]
```diff
@@ -56,3 +56,249 @@ spec:
         linstor.csi.linbit.com/on-storage-lost: remove # <--
     ...
 ```
+
+## Backup on S3 Storage
+
+> This feature requires a configured master passphrase (see instructions on the top of the page).
```
Comment (kvaps) [inline modules/031-linstor/docs/USAGE.md:62]:
```suggestion
> This feature requires a configured master passphrase (see instructions on the top of the [configuration](configuration.html#linstor-storage-configuration) page).
```

File: modules/031-linstor/docs/USAGE_RU.md [hunk 12]
```diff
@@ -54,3 +54,250 @@ spec:
         linstor.csi.linbit.com/on-storage-lost: remove # <--
     ...
 ```
+
+## Резервное копирование в S3
+
+> Использование данной возможности требует настроенного мастер-пароля (см. инструкции вначале страницы).
```
Comment (kvaps) [inline modules/031-linstor/docs/USAGE_RU.md:60]:
```suggestion
> Использование данной возможности требует настроенного мастер-пароля (см. инструкции вначале страницы [конфигурации](configuration.html#linstor-storage-configuration)).
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

File: modules/031-linstor/docs/USAGE_RU.md [hunk 13]
```diff
@@ -30,3 +30,250 @@ spec:
...
+      ```shell
+      kubectl exec -n d8-linstor deploy/linstor-controller -- linstor backup list <backup-remote-name>
+      ```
+
+      , где `<backup-remote-name>` — название backup-подключения, использованное в `VolumeSnapshotClass`.
+
+   1. Получите id-снапшота из имени объекта в S3-бакете через UI-интерфейс или CLI-утилиты S3-сервиса.
+
+1. Создайте `VolumeSnapshotContent`, указывающий на конкретный id снапшота.
+
```
Comment (kvaps) [inline modules/031-linstor/docs/USAGE_RU.md:219]:
```suggestion

   > VolumeSnapshotContent — ресурс на уровне кластера. Каждый VolumeSnapshotClass может быть связан только с одним VolumeSnapshot. Поэтому удостоверьтесь в уникальности его имени.

```

File: modules/031-linstor/docs/USAGE.md [hunk 14]
```diff
@@ -32,3 +32,249 @@ spec:
...
+      ```shell
+      kubectl exec -n d8-linstor deploy/linstor-controller -- linstor backup list <backup-remote-name>
+      ```
+
+      , where the `<backup-remote-name>` is the remote name used in the `VolumeSnapshotClass`.
+
+   1. Get a snapshot id from the object name in the S3 backet via the UI interface or CLI utilities of the S3 service.
+
+1. Create a `VolumeSnapshotContent`, pointing to a specific snapshot id.
+
```
Comment (kvaps) [inline modules/031-linstor/docs/USAGE.md:220]:
```suggestion

   > VolumeSnapshotContent is a cluster-wide resource. Each VolumeSnapshotClass can only be associated with one VolumeSnapshot. So make sure its name is unique.
   
```

File: modules/031-linstor/docs/USAGE.md [hunk 15]
```diff
@@ -32,3 +32,251 @@ spec:
...
+   > VolumeSnapshotContent is a cluster-wide resource. Each VolumeSnapshotClass can only be bound with one VolumeSnapshot. So make sure its name is unique.
+
```
Comment (z9r5) [inline modules/031-linstor/docs/USAGE.md:222]:
```suggestion

```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

**Review by kvaps — APPROVED** [review]
LGTM

**Pull request merged**